src/passes/MergeSimilarFunctions.cpp

```diff
@@ -246,8 +246,16 @@ bool MergeSimilarFunctions::areInEquvalentClass(Function* lhs,
       }
       auto* lhsCallee = module->getFunction(lhsCast->target);
       auto* rhsCallee = module->getFunction(rhsCast->target);
-      if (lhsCallee->getSig() != rhsCallee->getSig()) {
-        return false;
+      // In nominal typing we need the types to match perfectly. Otherwise,
+      // structurally is sufficient.
+      if (getTypeSystem() == TypeSystem::Nominal) {
+        if (lhsCallee->type != rhsCallee->type) {
+          return false;
+        }
```

Comment:
We should unconditionally compare the types for equality. Under equirecursive typing this will be equivalent to comparing the signature structures and for the other two type systems comparing the signatures is presumably wrong.

Reply:
Ah, thanks, I forgot how that worked in the structural case... done.